File: fbcore/src/main/java/com/facebook/common/streams/TailAppendingInputStream.java
/*
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 */

package com.facebook.common.streams;

import java.io.FilterInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.lang.Override;

/**
 * InputStream that returns all bytes from another stream, then appends the specified 'tail' bytes.
 */
public class TailAppendingInputStream extends FilterInputStream {
  private final byte[] mTail;
  private int mTailOffset;
  private int mMarkedTailOffset;

  public TailAppendingInputStream(InputStream inputStream, byte[] tail) {
    super(inputStream);
    if (inputStream == null) {
      throw new NullPointerException();
    }
    if (tail == null) {
      throw new NullPointerException();
    }
    mTail = tail;
  }

  @Override
  public int read() throws IOException {
    final int readResult = in.read();
    if (readResult != -1) {
      return readResult;
    }
    return readNextTailByte();
  }

  @Override
  public int read(byte[] buffer) throws IOException {
    return read(buffer, 0, buffer.length);
  }

  @Override
  public int read(byte[] buffer, int offset, int count) throws IOException {
    final int readResult = in.read(buffer, offset, count);
    if (readResult != -1) {
      return readResult;
    }

    if (count == 0) {
      return 0;
    }

    int bytesRead = 0;
    while (bytesRead < count) {
      final int nextByte = readNextTailByte();
      if (nextByte == -1) {
        break;
      }
      buffer[offset + bytesRead] = (byte) nextByte;
      bytesRead++;
    }
    return bytesRead > 0 ? bytesRead : -1;
  }

  @Override
  public void reset() throws IOException {
    if (in.markSupported()) {
      in.reset();
      mTailOffset = mMarkedTailOffset;
    } else {
      throw new IOException("mark is not supported");
    }
  }

  @Override
  public void mark(int readLimit) {
    if (in.markSupported()) {
      super.mark(readLimit);
      mMarkedTailOffset = mTailOffset;
    }
  }

  private int readNextTailByte() {
    if (mTailOffset >= mTail.length) {
      return -1;
    }
    return ((int) mTail[mTailOffset++]) & 0xFF;
  }
}

File: fbcore/src/main/java/com/facebook/common/streams/TailAppendingInputStream.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.common.streams

import java.io.FilterInputStream
import java.io.IOException
import java.io.InputStream

/**
 * InputStream that returns all bytes from another stream, then appends the specified 'tail' bytes.
 */
class TailAppendingInputStream(inputStream: InputStream, tail: ByteArray) :
    FilterInputStream(inputStream) {

  private val tail: ByteArray
  private var tailOffset = 0
  private var markedTailOffset = 0

  init {
    if (inputStream == null) {
      throw NullPointerException()
    }
    if (tail == null) {
      throw NullPointerException()
    }
    this.tail = tail
  }

  @Throws(IOException::class)
  override fun read(): Int {
    val readResult = `in`.read()
    if (readResult != -1) {
      return readResult
    }
    return readNextTailByte()
  }

  @Throws(IOException::class)
  override fun read(buffer: ByteArray): Int = read(buffer, 0, buffer.size)

  @Throws(IOException::class)
  override fun read(buffer: ByteArray, offset: Int, count: Int): Int {
    val readResult = `in`.read(buffer, offset, count)
    if (readResult != -1) {
      return readResult
    }

    if (count == 0) {
      return 0
    }

    var bytesRead = 0
    while (bytesRead < count) {
      val nextByte = readNextTailByte()
      if (nextByte == -1) {
        break
      }
      buffer[offset + bytesRead] = nextByte.toByte()
      bytesRead++
    }
    return if (bytesRead > 0) bytesRead else -1
  }

  @Throws(IOException::class)
  override fun reset() {
    if (`in`.markSupported()) {
      `in`.reset()
      tailOffset = markedTailOffset
    } else {
      throw IOException("mark is not supported")
    }
  }

  override fun mark(readLimit: Int) {
    if (`in`.markSupported()) {
      super.mark(readLimit)
      markedTailOffset = tailOffset
    }
  }

  private fun readNextTailByte(): Int {
    if (tailOffset >= tail.size) {
      return -1
    }
    return (tail[tailOffset++].toInt()) and 0xFF
  }
}
